scripts/install-cni.sh

```diff
@@ -121,15 +121,24 @@ if [ "$ENABLE_MASQUERADE" == "true" ]; then
   fi
 fi
 
-if [ "$ENABLE_PRIVATE_IPV6_ACCESS" == "true" ]; then
+if [ "$ENABLE_PRIVATE_IPV6_ACCESS" == "true" ] || [ "$ENABLE_IPV6" == "true" ]; then
   node_ipv6_addr=$(curl -s -k --fail "http://metadata.google.internal/computeMetadata/v1/instance/network-interfaces/0/?recursive=true" -H "Metadata-Flavor: Google" | jq -r '.ipv6s[0]' ) ||:
 
   if [ -n "${node_ipv6_addr:-}" ] && [ "${node_ipv6_addr}" != "null" ]; then
     echo "Found nic0 IPv6 address ${node_ipv6_addr:-}. Filling IPv6 subnet and route..."
-    cni_spec=$(echo ${cni_spec:-} | sed -e \
-      "s#@ipv6SubnetOptional#, [{\"subnet\": \"${node_ipv6_addr:-}/112\"}]#g;
-       s#@ipv6RouteOptional#, {\"dst\": \"::/0\"}#g")
-    
+
+    if [ "$ENABLE_IPV6" == "true" ]; then
+      ipv6_subnet=$(echo $response | jq '.spec.podCIDRs[1]')
```

Comment:
Which component is responsible to fill in `.spec.podCIDRs[1]`? Should the script exit and retry if this is not there yet?

Reply:
I am working on the k8s changes required to populate .spec.podCIDRs[1].

The changes will be in:
https://github.com/kubernetes/legacy-cloud-providers/blob/master/gce/gce_instances.go#L495
and 
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go#L284

With those changes, we should always have a value in .spec.podCIDRs[1] when IPV6/DualStack is enabled.

Reply:
Basant I was testing this and it seems like there is a race in the GCE metadata server and populating the interface IP. Basically what we end up writing in the podCIDRs is sometimes wrong and stale. For now we can continue using the interface's IP address to write the CNI config just like you had suggested. This would look like this:

s#@ipv6SubnetOptional#, [{\"subnet\": \"${node_ipv6_addr%:*:*}:ffff:0/112\"}]#g;

If it is safe we should also do the same thing for directpath. We cannot use the first /112 because then we will conflict with the node IP. When dual stack clusters are enabled, we want to use the last /112 for pods.

Reply:
@sdmodi I'll make this change and test. Thanks

Reply:
Done. Made the change and tested only for IPV6 Dual Stack for now. 
@varunmar @MrHohn do you see any issues with changing directpath to use last /112 block instead of first /112 block?

Reply:
Not an expert on the directpath usage with pod - will defer to @varunmar to respond :)

Reply:
I'm not sure  I understand the question - the first/last 112 block of the node allocation? How big is the range handed out to the node - it used to only be /112, is it bigger now?

Reply:
Node is assigned a /96 block. Node uses the first address in this block for its interface. So we need to pick a /112 block from it. I am picking the last block to avoid any ip conflicts. Want to check if this applies for directpath too.

Reply:
Made it consistent with directpath. Since the node address is :0 and the first address allocated to pods starts from :1, we should not see any conflicts.